probe-rs-tools/src/bin/probe-rs/util/cli.rs

```diff
@@ -519,53 +637,81 @@ async fn print_monitor_event(
+            let channel = channel as usize;
+            let Some(processor) = client.channel_processors.get_mut(channel) else {
                 return;
             };
 
-            processor.process(&bytes);
-        }
-        MonitorEvent::SemihostingOutput(SemihostingOutput::StdOut(str)) => {
-            print!("{}", str)
+            processor
+                .process(
+                    &bytes,
+                    // See ChannelIdentifier on why we access with clones here; also, while it'd be
+                    // more efficient to resolve those lookups at channel discovery, it doesn't really
+                    // matter, and again, ease of maintenance beats theoretical performance unless
+                    // benchmarked otherwise.
+                    ChannelIdentifier::Rtt(processor.channel.clone()).find_in(target_output_files),
+                )
+                .await;
         }
-        MonitorEvent::SemihostingOutput(SemihostingOutput::StdErr(str)) => {
-            eprint!("{}", str)
+        MonitorEvent::SemihostingOutput(semihosting) => {
+            let (channel, written) = match semihosting {
+                SemihostingOutput::StdOut(str) => {
+                    print!("{}", str);
+                    (ChannelIdentifier::Semihosting("stdout".to_string()), str)
+                }
+                SemihostingOutput::StdErr(str) => {
+                    eprint!("{}", str);
+                    (ChannelIdentifier::Semihosting("stderr".to_string()), str)
+                }
+            };
+
+            if let Some(remote_processor) = channel.find_in(target_output_files) {
+                // Silently discarding output file errors
+                _ = remote_processor.write_all(written.as_bytes()).await;
+            };
         }
     }
 }
 
 struct Channel {
     channel: String,
     decoder: RttDecoder,
-    printer: Printer,
+    printer_prefix: String,
 }
 
 impl Channel {
     fn new(channel: String, decoder: RttDecoder) -> Self {
         Self {
             channel,
             decoder,
-            printer: Printer {
-                prefix: String::new(),
-            },
+            printer_prefix: String::new(),
         }
     }
 
     fn print_channel_name(&mut self, width: usize) {
-        self.printer.prefix = format!("[{:width$}] ", self.channel, width = width);
+        self.printer_prefix = format!("[{:width$}] ", self.channel, width = width);
     }
 
-    fn process(&mut self, bytes: &[u8]) {
-        _ = self.decoder.process(bytes, &mut self.printer);
+    async fn process(&mut self, bytes: &[u8], copy_to: Option<&mut tokio::fs::File>) {
+        let mut printer = Printer {
+            prefix: &self.printer_prefix,
+            copy_to,
+        };
+        let _ = self.decoder.process(bytes, &mut printer).await;
     }
 }
 
-struct Printer {
-    prefix: String,
+struct Printer<'a> {
+    prefix: &'a str,
+    copy_to: Option<&'a mut tokio::fs::File>,
```

Comment:
This is okay, but it would for sure be better if we could somehow compose different handlers. The printer printring to a file seems slightly off. No matter, it's fine in the binary.

Reply:
It should be possible to turn RttDataHandler into something that can be chained, but I think that as long as the collectors are as simple as they are now, that'd be a bit excessive.